rust/src/adt.rs

```diff
@@ -0,0 +1,846 @@
+// SPDX-License-Identifier: MIT
+use core::ffi::*;
+use core::mem::size_of;
+
+use crate::c_size_t;
+
+#[derive(Debug)]
+pub enum AdtError {
+    NotFound = -1,
+    BadOffset = -4,
+    BadPath = -5,
+    BadNCells = -14,
+    BadValue = -15,
+    BadLength = -20,
+}
+
+const ADT_ALIGN: i32 = 4;
+
+#[repr(C)]
```

Comment:
Fat pointers like &[T] do not have a stable C representation despite rust allowing you to put #[repr(C)] on a struct containing one.

Reply:
Better to declare it as `*const u8` then? The struct doesn't actually do anything, it was a leftover from a previous architecture. Probably fine to just get rid of it and move the functions in the ADT impl out of it.